app/src/rx_fastmetadata_app.c

```diff
@@ -4,12 +4,152 @@
 
 #include "rx_fastmetadata_app.h"
 
-static void app_rx_fmd_handle_rtp(struct st_app_rx_fmd_session* s) {
-  dbg("%s(%d).\n", __func__, s->idx);
+static void app_rx_fmd_close_source(struct st_app_rx_fmd_session* session) {
+  if (session->st41_ref_fd >= 0) {
+    munmap(session->st41_ref_begin, session->st41_ref_end - session->st41_ref_begin);
+    close(session->st41_ref_fd);
+    session->st41_ref_fd = -1;
+  }
+}
+
+static int app_rx_fmd_open_ref(struct st_app_rx_fmd_session* session) {
+  int fd, idx = session->idx;
+  struct stat i;
+  session->st41_ref_fd = -1;
+
+  fd = st_open(session->st41_ref_url, O_RDONLY);
+  if (fd < 0) {
+    info("%s(%d), open file '%s' fail.\n", __func__, idx, session->st41_ref_url);
+    return -EIO;
+  }
+
+  if (fstat(fd, &i) < 0) {
+    err("%s, fstat '%s' fail.\n", __func__, session->st41_ref_url);
+    close(fd);
+    return -EIO;
+  }
+
+  uint8_t* m = mmap(NULL, i.st_size, PROT_READ, MAP_SHARED, fd, 0);
+  if (MAP_FAILED == m) {
+    err("%s(%d), mmap '%s' fail.\n", __func__, idx, session->st41_ref_url);
+    close(fd);
+    return -EIO;
+  }
+
+  session->st41_ref_begin = m;
+  session->st41_ref_cursor = m;
+  session->st41_ref_end = m + i.st_size;
+  session->st41_ref_fd = fd;
+
+  info("%s, opening file '%s' success.\n", __func__, session->st41_ref_url);
+  return 0;
+}
+
+static int app_rx_fmd_compare_with_ref(struct st_app_rx_fmd_session* session, void* frame,
+                                       int frame_size) {
+  int ret = -1;
+  uint32_t last_zeros = 0; /* 4 bytes with 0 */
+  uint32_t st41_ref_remaining_length =
+      session ? session->st41_ref_end - session->st41_ref_cursor : 0;
+
+  if (frame_size <= st41_ref_remaining_length) {
+    ret = memcmp(frame, session->st41_ref_cursor, frame_size);
+    if (ret) {
+      session->errors_count++;
+      err("%s() FAIL: reference file comparison (1).\n", __func__);
```

Comment:
this seems a litle bit confusing to me 
why are we using stage 1 stage 2 stage 3 in logs ? 
can't we include anything more specific ? 

reference file comparison inframe memory differences ?
chunk of memory is different ?

I belive this would make the code more clear and the messages understandable at first glance instead of making anyone testing that look into code what 1 / 2 / 3 means

Reply:
That makes it much better imo